### PR TITLE
nixos/matrix-synapse: protect created files

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change

Enforce UMask on the systemd unit to restrict the permissions of files
created. Especially the homeserver signing key should not be world
readable, and media is served through synapse itself, so no other user
needs access to these files.

Use a prestart chmod to fixup the permissions on the signing key.

Discovered by @andir 

```
# ls -lah /var/lib/matrix-synapse/
total 4.0K
drwx------ 1 matrix-synapse matrix-synapse  54 May  8  2020 .
drwxr-xr-x 1 root           root           276 May 13 06:15 ..
-rw-r--r-- 1 matrix-synapse matrix-synapse  59 May  8  2020 homeserver.signing.key
drwxr-xr-x 1 matrix-synapse matrix-synapse 176 May 17  2020 media
```

**Not critical, since the StateDirectory is 0700.**

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/misc/matrix-synapse.nix [hunk 1]
```diff
@@ -719,6 +719,9 @@ in {
           --config-path ${configFile} \
           --keys-directory ${cfg.dataDir} \
           --generate-keys
+
+        # restrict permissions on already existing signing keys
+        chmod 0600 ${cfg.dataDir}/homeserver.signing.key
```
Comment (mweinelt) [inline nixos/modules/services/misc/matrix-synapse.nix:724]: Maybe also on the media directory?

Comment (andir) [reply nixos/modules/services/misc/matrix-synapse.nix:724]: Is it actually possible to serve the media without going through synapse? That would be very reasonable IMHO but unlikely as synapse will fetch them lazily from other servers if they are missing from the local copy.

Comment (mweinelt) [reply nixos/modules/services/misc/matrix-synapse.nix:724]: They might also be encrypted, so I think they should be served through synapse.

Comment (andir) [reply nixos/modules/services/misc/matrix-synapse.nix:724]: Why? It is supposed to be end-to-end encryption. The server shouldn't even be able to decrypt them.

Comment (mweinelt) [reply nixos/modules/services/misc/matrix-synapse.nix:724]: I wonder if it encapsulates the message in some form; though I only see `blob:` scheme urls for images in element.

Comment (ajs124) [reply nixos/modules/services/misc/matrix-synapse.nix:724]: There is https://github.com/turt2live/matrix-media-repo, so it's possible somehow, if I'm not misunderstanding things.

Comment (mweinelt) [reply nixos/modules/services/misc/matrix-synapse.nix:724]: Though that looks like a scenario where the media history is shared among multiple homeservers or even outsourced to an S3 backend. Either way we could set the media directory 750 and thereby require group access to read and access it. Does that sound good?

**Review by andir — APPROVED** [review]
Regardless of the media situation this is already better. Tests run for me.

**Comment by Ma27** [comment]
Deploying this to an existing `matrix-synapse` causes the following error:

```
May 22 17:16:31 matrix matrix-synapse-pre-start[5446]: chmod: changing permissions of '/var/lib/matrix-synapse/homeserver.signing.key': Operation not permitted
```

**Comment by mweinelt** [comment]
@Ma27 Who owns your signing key, if not the `matrix-synapse` user that runs the unit?

**Comment by Ma27** [comment]
@mweinelt `root`:

```
[root@matrix:/var/lib/matrix-synapse]# ls -lah homeserver.signing.key
-rw-r--r-- 1 root root 59 Mar  3  2019 homeserver.signing.key
```

I guess this is because the `matrix-synapse` module had `PermissionsStartOnly = true` until Dec 2019 and the key is generated in `ExecStartPre`.

**Comment by mweinelt** [comment]
I've added a separate, privileged prestart command that should cover this legacy situation.

**Review by Ma27 — APPROVED** [review]
LGTM for master. Backport will be a little more work as we'll need https://github.com/NixOS/nixpkgs/pull/109976 for a successful evaluation. Will prepare a PR for that.

**Pull request merged**